File: go/dsa/data-structures/stack-queue/min-stack.go
type MinStack struct {
    data    []int
    minData []int
}

func Constructor() MinStack { // @viz:init_stack
    return MinStack{ // @viz:init_min
        data:    []int{},
        minData: []int{},
    }
}

func (this *MinStack) Push(val int) { // @viz:push_start
    this.data = append(this.data, val) // @viz:push_val
    if len(this.minData) == 0 || val <= this.minData[len(this.minData)-1] { // @viz:min_check
        this.minData = append(this.minData, val) // @viz:min_push
    }
} // @viz:push_end

func (this *MinStack) Pop() { // @viz:pop_start
    popped := this.data[len(this.data)-1] // @viz:pop_val
    this.data = this.data[:len(this.data)-1]
    if popped == this.minData[len(this.minData)-1] { // @viz:pop_min_check
        this.minData = this.minData[:len(this.minData)-1] // @viz:min_pop
    }
} // @viz:pop_end

func (this *MinStack) Top() int { // @viz:top_start
    return this.data[len(this.data)-1] // @viz:top_val
}

func (this *MinStack) GetMin() int { // @viz:get_min_start
    return this.minData[len(this.minData)-1] // @viz:min_val
}

File: javascript/dsa/data-structures/stack-queue/min-stack.js
class MinStack {
  constructor() { // @viz:init_stack
    this.values = [];
    this.minimums = []; // @viz:init_min
  }

  push(val) { // @viz:push_start
    this.values.push(val); // @viz:push_val
    if (this.minimums.length === 0 || val <= this.minimums[this.minimums.length - 1]) { // @viz:min_check
      this.minimums.push(val); // @viz:min_push
    }
  } // @viz:push_end

  pop() { // @viz:pop_start
    const popped = this.values.pop(); // @viz:pop_val
    if (popped === this.minimums[this.minimums.length - 1]) { // @viz:pop_min_check
      this.minimums.pop(); // @viz:min_pop
    }
  } // @viz:pop_end

  top() { // @viz:top_start
    return this.values[this.values.length - 1]; // @viz:top_val
  }

  getMin() { // @viz:get_min_start
    return this.minimums[this.minimums.length - 1]; // @viz:min_val
  }
}